src/jit/lowerxarch.cpp

```diff
@@ -3511,6 +3523,20 @@ void Lowering::TreeNodeInfoInitCmp(GenTreePtr tree)
             tree->gtFlags |= GTF_UNSIGNED;
         }
     }
+
+    // If op1 is of TYP_BYTE or TYP_UBYTE, we need to make sure it is constrained to using
+    // only byteable registers
+    if (varTypeIsByte(op1))
+    {
+        regMaskTP regMask;
+        bool containedNode = (op1->gtLsraInfo.srcCount == 0) && (op1->gtLsraInfo.dstCount == 0);
+        if (!containedNode)
+        {
+            regMask = op1->gtLsraInfo.getSrcCandidates(m_lsra);
+            assert(regMask != RBM_NONE);
+            op1->gtLsraInfo.setSrcCandidates(m_lsra, regMask & ~RBM_NON_BYTE_REGS);
+        }
```

Comment:
A compare node that produces a result in a register requires its destination candidates be a byteable reg because it uses "setcc  bytereg" instruction.